Say missing key

maettu opened the pull request:
Tell the user which mandatory property is missing.
The code is not elegant, though :-(
(And I'd like to learn how to write that shorter and more elegantly.)

maettu committed:
Say missing key

Tell the user which mandatory property is missing.
The code is not elegant, though :-(

Altai-man commented on lib/JSON/Schema.pm6:
```diff
@@ -375,9 +375,13 @@ class JSON::Schema {
-                die X::JSON::Schema::Failed.new:
-                    :$!path, :reason("Object does not have required property");
+            if $value ~~ Associative {
+                for @.prop -> $mandatory_prop {
```
Please replace `@.prop` with `@!prop`, it'll be slightly faster this way(in ideal world it'd be optimized by the compiler to be the same, but not right now).

Altai-man replied:
`@.prop` is actually a call to `self.prop` attribute getter, while `@!prop` compiles into direct access to the attribute.

jnthn replied:
@Altai-man By this point the JIT does pretty good on inlining accessors, so they come out pretty alright. :-)

Altai-man replied:
Thanks, glad to update my knowledge.

maettu replied:
For readability I believe @.prop to be better. That's how it is is declared a few lines earlier.

jnthn replied:
When reading a usage of `@.prop`, since it's actually a method call, it's possible that any subclass of the current class might override the accessor method. By contrast, when reading a usage of `@!prop`, you know it's referring unambiguously to the bit of storage declared in this class. Thus I'd argue `@!prop` is more readable in that you can read it and know precisely what it will do, while with `@.prop` there's potential polymorphism in there.

Altai-man commented on lib/JSON/Schema.pm6:
```diff
@@ -375,9 +375,13 @@ class JSON::Schema {
+                        die X::JSON::Schema::Failed.new:
+                            :$!path, :reason("Missing required property '$mandatory_prop'");
+                    }
+                }
```
Otherwise I do not see if as not elegant in any way. It is, in fact, much better than it was, because for one-liner I have sacrificed some clarity in this place. Your code is much more readable for me.

jnthn replied:
I think I'd probably write it as `if @!prop.grep({ not $value{$_}.defined }) -> @missing { ... }` and then make the reason be `"Missing required properties: @missing.map({ qq/'$_'/ }).join(', ')"`, to provide all of the names up front.

jnthn commented on lib/JSON/Schema.pm6:
```diff
@@ -375,9 +375,13 @@ class JSON::Schema {
-                die X::JSON::Schema::Failed.new:
-                    :$!path, :reason("Object does not have required property");
+            if $value ~~ Associative {
+                for @.prop -> $mandatory_prop {
```
The declaration `has Str @.prop` is short for `has Str @!prop; method prop() { @!prop }`. A usage of `@.prop` is equivalent to `$(self.prop)` - that is, a (polymorphic) method call. Most of the time inside of a class, we simply want the attribute, so I'd retain the `@!prop` here.

jnthn replied:
The naming convention usually followed in Perl 6 is to use `-` rather than `_` (not by everyone, but it's done in this module, and local consistency is nice).

jnthn reviewed (approved):
Ah, I see now I duped feedback a bit form @Altai-man (opened this window, went away for a while, came back and reviewed). :-)

maettu committed:
improve, make more elegant

Thanks @jnthn and @Altai-man.
Use grep-map instead of for loop.
Catch all missing keys at once.
(Bonus: The non-conventional variable went away.)
